clouddriver-sql/src/main/kotlin/com/netflix/spinnaker/clouddriver/sql/event/SqlEventRepository.kt

```diff
@@ -0,0 +1,298 @@
+import de.huxhorn.sulky.ulid.ULID
+import io.github.resilience4j.retry.Retry
+import io.github.resilience4j.retry.RetryConfig
+import io.github.resilience4j.retry.RetryRegistry
+import org.jooq.Condition
+import org.jooq.DSLContext
+import org.jooq.impl.DSL.currentTimestamp
+import org.jooq.impl.DSL.field
+import org.jooq.impl.DSL.table
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.validation.Validator
+import java.time.Duration
+import java.util.Optional
+import java.util.UUID
+
+class SqlEventRepository(
+  private val jooq: DSLContext,
+  sqlProperties: SqlProperties,
+  private val serviceVersion: ServiceVersion,
+  private val objectMapper: ObjectMapper,
+  private val applicationEventPublisher: ApplicationEventPublisher,
+  private val registry: Registry,
+  private val validator: Validator,
+  private val retryRegistry: RetryRegistry
+) : EventRepository {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+  private val retryProperties: SqlRetryProperties = sqlProperties.retries
+
+  private val eventCountId = registry.createId("eventing.events")
+
+  /**
+   * TODO(rz): Performance.
+   */
+  override fun save(
+    aggregateType: String,
+    aggregateId: String,
+    originatingVersion: Long,
+    newEvents: List<SpinnakerEvent>
+  ) {
+    val eventNames = newEvents.joinToString { it.javaClass.simpleName }
+    log.debug("Saving $aggregateType/$aggregateId expecting version $originatingVersion with [$eventNames]")
+
+    val aggregateCondition = field("aggregate_type").eq(aggregateType)
+      .and(field("aggregate_id").eq(aggregateId))
+
+    // TODO(rz): Get this from Spring?
+    val retry = RetryConfig.custom<Retry>()
+      .maxAttempts(retryProperties.transactions.maxRetries)
+      .waitDuration(Duration.ofMillis(retryProperties.transactions.backoffMs))
+      .ignoreExceptions(AggregateChangeRejectedException::class.java)
+      .build()
+
+    try {
+      withPool(POOL_NAME) {
+        jooq.transactional(retryRegistry.retry("eventSave", retry)) { ctx ->
+          // Get or create the aggregate and immediately assert that this save operation is being committed against the
+          // most recent aggregate state.
+          val aggregate = ctx.maybeGetAggregate(aggregateCondition) ?: {
+            if (originatingVersion != 0L) {
+              // The aggregate doesn't exist and we're already expecting a non-zero version.
+              throw AggregateChangeRejectedException(-1, originatingVersion)
+            }
+
+            // The aggregate doesn't exist yet, so we'll go ahead and seed it immediately.
+            // Aggregate tokens are globally unique across all aggregates and calculated monotonically.
+            val initialAggregate = mapOf(
+              field("aggregate_type") to aggregateType,
+              field("aggregate_id") to aggregateId,
+              field("token") to getNextMonotonicToken(ctx),
```

Comment:
What value does reading from the DB via `getNextMonotonicToken` provides? What prevents collisions if 2 instances enter that in the same millisecond? Is that worth having to rerun the transactional closure?

Reply:
Removed after IRL convo; no longer getting monotomic.